### PR TITLE
Automatic Shield Management

### PR DESCRIPTION
**Pull request by Karthas077**
## Additions

- Shields should automatically be removed from apparel and placed into inventory when a weapon is equipped which cannot be used at the same time.
- Shields in inventory should automatically be worn if a weapon which allows their used is equipped.
- Intercepts the outfit manager's attempts to wear a shield while one is desired by loadout (Lets the loadout manager pick it up)
- Intercepts the outfit manager's attempts to wear a shield belt while a projectile weapon is desired by loadout.
- Adds an inventory right-click menu to put away a worn shield or equip a held shield.

## Reasoning

- Shields can already be dropped and picked up and/or dropped and worn to have a pawn alternate between two handed weapon and single-handed + shield.
- Weapons can already be put away and equipped with single clicks from the gear management pane.
- Increases compatibility with any mods which cause pawns to swap between different weapons by automating the process in addition to allowing it to be done manually.

## Alternatives

Honestly didn't really consider any alternatives. Only including the manual swapping maybe?

## Testing

Check tests you have performed:
- [x] Compiles without warnings
- [x] Game runs without errors
- [x] (For compatibility patches) ...with and without patched mod loaded
- [x] Playtested a colony (specify how long)

### CODE REVIEW

File: Source/CombatExtended/CombatExtended/Loadouts/ITab_Inventory.cs [hunk 1]
```diff
@@ -349,8 +349,36 @@ private void DrawThingRow(ref float y, float width, Thing thing, bool showDropBu
                             floatOptionList.Add(equipOption);
                         }
                     }
+                    //Thing we're hovering over is a shield
+                    if (thing is Apparel_Shield)
```
Comment (perkinslr) [inline Source/CombatExtended/CombatExtended/Loadouts/ITab_Inventory.cs:353]: No reason to check this then cast to Apparel below, we can use C#7 features :smile: 

`if (thing is Apparel_Shield shield)`

File: Source/CombatExtended/Harmony/Harmony_JobGiver_OptimizeApparel.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+using Verse;
+using Verse.AI;
+
+namespace CombatExtended.HarmonyCE
+{
+    /* Overrides the CompReloadable reload job if the pawn has suitable ammo in their inventory.
```
Comment (perkinslr) [inline Source/CombatExtended/Harmony/Harmony_JobGiver_OptimizeApparel.cs:12]: You copied this patch class from another file but didn't update the description.

File: Source/CombatExtended/CombatExtended/Loadouts/ITab_Inventory.cs (hunk 1 above)
Comment (Karthas077) [reply Source/CombatExtended/CombatExtended/Loadouts/ITab_Inventory.cs:353]: I have no idea what that means XD If you suggest a code change that uses those I will implement/accept them

**Comment by Tostov** [comment]
How automated is the auto equip/unequip?

While the behavior is generally desirable, I can imagine instances where, combined with the new features, it might result in undersirable behavior, especially if equipping from inventory takes time. For example, stowing a shield when using a one handed weapon so the shield isn't completely destroyed by enemy attacks, or force equipping a shield while using a two handed weapon in a crisis situation.

**Comment by Karthas077** [comment]
> How automated is the auto equip/unequip?
> 
> While the behavior is generally desirable, I can imagine instances where, combined with the new features, it might result in undersirable behavior, especially if equipping from inventory takes time. For example, stowing a shield when using a one handed weapon so the shield isn't completely destroyed by enemy attacks, or force equipping a shield while using a two handed weapon in a crisis situation.

Just like how swapping between weapons is currently free/instant, all of the automatic shield swaps are free/instant.

The actual triggers are ON EQUIP, so you can 100% un-equip a shield and the automated system will not re-equip it until you either equip it manually, or equip a new 1 handed weapon.

As for equipping a shield while using a two-handed weapon, if you have any valid 1-handed weapon in your inventory it will equip that, otherwise it will leave your two-handed weapon equipped.

**Comment by N7Huntsman** [comment]
While it will stow the shield if you equip the shield and *then* the two-handed weapon, it won't stow the weapon when you equip the weapon first.

File: Source/CombatExtended/CombatExtended/Loadouts/Loadout.cs [hunk 3]
```diff
@@ -173,6 +180,14 @@ public void MoveSlot(LoadoutSlot slot, int toIndex)
         public void RemoveSlot(LoadoutSlot slot)
         {
             _slots.Remove(slot);
+            if (containsShield && slot.thingDef.thingCategories.Contains(CE_ThingCategoryDefOf.Shields))
+            {
+                //slot being removed was for a shield. Check if any other shields remain
+                if (Slots.FirstOrDefault(s => s.count >= 1 && s.thingDef.thingCategories.Contains(CE_ThingCategoryDefOf.Shields)) == null )
```
Comment (LockdownX7) [inline Source/CombatExtended/CombatExtended/Loadouts/Loadout.cs:186]: This causes an error when removing items from a loadout:
```
Exception filling window for CombatExtended.Dialog_ManageLoadouts: System.NullReferenceException: Object reference not set to an instance of an object
  at CombatExtended.Loadout.RemoveSlot (CombatExtended.LoadoutSlot slot) [0x0001c] in <7bb6722d09a24bc0aed468d7b5493df0>:0 
  at CombatExtended.Dialog_ManageLoadouts.DrawSlot (UnityEngine.Rect row, CombatExtended.LoadoutSlot slot, System.Boolean slotDraggable) [0x004dc] in <7bb6722d09a24bc0aed468d7b5493df0>:0 
  at CombatExtended.Dialog_ManageLoadouts.DrawSlotList (UnityEngine.Rect canvas) [0x00203] in <7bb6722d09a24bc0aed468d7b5493df0>:0 
  at CombatExtended.Dialog_ManageLoadouts.DoWindowContents (UnityEngine.Rect canvas) [0x0051c] in <7bb6722d09a24bc0aed468d7b5493df0>:0 
  at Verse.Window.InnerWindowOnGUI (System.Int32 x) [0x00165] in <c36f9493c9844ddaa7fb5c788416098f>:0 
Verse.Log:Error(String, Boolean)
Verse.Window:InnerWindowOnGUI(Int32)
UnityEngine.GUI:CallWindowDelegate(WindowFunction, Int32, Int32, GUISkin, Int32, Single, Single, GUIStyle)
```

File: Source/CombatExtended/Harmony/Harmony_JobGiver_OptimizeApparel.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+namespace CombatExtended.HarmonyCE
+{
+    /* Intercepts the outfit manager's attempts to equip a shield or shield belt.
+     * If the current loadout has a projectile weapon in it, does not let the outfit manager grab a shield belt.
+     * If the current loadout has a shield in it, does not let the outfit manager grab a shield.
+     */
+    [HarmonyPatch(typeof(JobGiver_OptimizeApparel), "ApparelScoreGain_NewTmp")]
+    internal static class Harmony_JobGiver_OptimizeApparel_ApparelScoreGain_NewTmp
+    {
+        internal static bool Prefix(Pawn pawn, Apparel ap, ref float __result)
```
Comment (LockdownX7) [inline Source/CombatExtended/Harmony/Harmony_JobGiver_OptimizeApparel.cs:19]: Shouldn't this patch also check if the pawn is equipped with a two-handed weapon, and return -1000f if the item is Apparel_Shield? Right now colonists with a two-handed weapon will still go grab a shield if their outfit allows it, requiring manual intervention by the player.

**Review by LockdownX7 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**